mirrord/cli/src/port_forward.rs

```diff
@@ -55,44 +56,82 @@ pub struct PortForwarder {
     ping_pong_timeout: Instant,
 }
 
+#[derive(Clone, Debug, PartialEq)]
+pub struct ResolvedPortMapping {
+    pub local: SocketAddr,
+    pub remote: SocketAddr,
+}
+
 /// Used by tasks for individual forwarding connections to send instructions to [`PortForwarder`]'s
 /// main loop.
 #[derive(Debug)]
 enum PortForwardMessage {
     /// A request to make outgoing connection to the remote peer.
     /// Sent by the task only after receiving first batch of data from the user.
     /// The task waits for [`ConnectionId`] on the other end of the [`oneshot`] channel.
-    Connect(PortMapping, oneshot::Sender<ConnectionId>),
+    Connect(ResolvedPortMapping, oneshot::Sender<ConnectionId>),
 
     /// Data received from the user in the connection with the given id.
     Send(ConnectionId, Vec<u8>),
 
     /// A request to close the remote connection with the given id, if it exists.
-    Close(PortMapping, Option<ConnectionId>),
+    Close(ResolvedPortMapping, Option<ConnectionId>),
 }
 
 impl PortForwarder {
     pub(crate) async fn new(
-        agent_connection: AgentConnection,
+        mut agent_connection: AgentConnection,
         parsed_mappings: Vec<PortMapping>,
     ) -> Result<Self, PortForwardError> {
         // open tcp listener for local addrs
         let mut listeners = StreamMap::with_capacity(parsed_mappings.len());
-        let mut mappings = HashMap::with_capacity(parsed_mappings.len());
+        let mut mappings: HashMap<SocketAddr, SocketAddr> =
+            HashMap::with_capacity(parsed_mappings.len());
 
         if parsed_mappings.is_empty() {
             return Err(PortForwardError::NoMappingsError());
         }
+
+        // setup agent connection
+        agent_connection
+            .sender
+            .send(ClientMessage::SwitchProtocolVersion(
+                mirrord_protocol::VERSION.clone(),
+            ))
+            .await?;
+        match agent_connection.receiver.recv().await {
+            Some(DaemonMessage::SwitchProtocolVersionResponse(version))
+                if CLIENT_READY_FOR_LOGS.matches(&version) =>
+            {
+                agent_connection
+                    .sender
+                    .send(ClientMessage::ReadyForLogs)
+                    .await?;
+            }
+            _ => return Err(PortForwardError::AgentConnectionFailed),
+        }
+
         for mapping in parsed_mappings {
             if listeners.contains_key(&mapping.local) {
                 // two mappings shared a key thus keys were not unique
                 return Err(PortForwardError::PortMapSetupError(mapping.local));
             }
+            let resolved_remote = match mapping.remote.0 {
+                RemoteAddr::Ip(ip) => IpAddr::V4(ip),
+                RemoteAddr::Hostname(hostname) => {
+                    PortForwarder::resolve_hostname(&mut agent_connection, hostname)
```

Comment:
I think we want to resolve the hostname lazily (only right before making remote connection via agent). Probably requires another variant of `PortForwardMessage`